packages/component-base/src/data-provider-controller/cache.js

```diff
@@ -0,0 +1,187 @@
+   */
+  constructor(controller, pageSize, size, parentCache, parentCacheIndex) {
+    this.controller = controller;
+    this.pageSize = pageSize;
+    this.size = size || 0;
+    this.effectiveSize = size || 0;
+    this.parentCache = parentCache;
+    this.parentCacheIndex = parentCacheIndex;
+  }
+
+  /**
+   * An item in the parent cache that the current cache is associated with.
+   *
+   * @return {object | undefined}
+   */
+  get parentItem() {
+    return this.parentCache && this.parentCache.items[this.parentCacheIndex];
+  }
+
+  /**
+   * An array of sub-caches sorted in the same order as their associated items
+   * appear in the `items` array.
+   *
+   * @return {Cache[]}
+   */
+  get subCaches() {
+    return Object.values(this.#subCacheByIndex);
+  }
+
+  /**
+   * Whether the cache or any of its descendant caches have pending requests.
+   *
+   * @return {boolean}
+   */
+  get isLoading() {
+    if (this.pendingRequests.size > 0) {
+      return true;
+    }
+
+    return this.subCaches.some((subCache) => subCache.isLoading);
+  }
+
+  /**
+   * Recalculates the effective size for the cache and its descendant caches recursively.
+   */
+  recalculateEffectiveSize() {
+    this.effectiveSize =
+      !this.parentItem || this.controller.isExpanded(this.parentItem)
+        ? this.size +
+          this.subCaches.reduce((total, subCache) => {
+            subCache.recalculateEffectiveSize();
+            return total + subCache.effectiveSize;
+          }, 0)
+        : 0;
+  }
+
+  /**
+   * Adds an array of items corresponding to the given page
+   * to the `items` array.
+   *
+   * @param {number} page
+   * @param {object[]} items
+   */
+  setPage(page, items) {
+    const startIndex = page * this.pageSize;
+    items.forEach((item, i) => {
+      this.items[startIndex + i] = item;
+    });
+  }
+
+  /**
+   * Retrieves the sub-cache associated with the item at the given index
+   * in the `items` array.
+   *
+   * @param {number} index
+   * @return {Cache | undefined}
+   */
+  getSubCache(index) {
+    return this.#subCacheByIndex[index];
+  }
+
+  /**
+   * Removes the sub-cache associated with the item at the given index
+   * in the `items` array.
+   *
+   * @param {number} index
+   */
+  removeSubCache(index) {
+    delete this.#subCacheByIndex[index];
+  }
+
+  /**
+   * Removes all sub-caches.
+   */
+  removeSubCaches() {
+    this.#subCacheByIndex = {};
+  }
+
+  /**
+   * Creates and associates a sub-cache for the item at the given index
+   * in the `items` array.
+   *
+   * @param {number} index
+   * @return {Cache}
+   */
+  createSubCache(index) {
+    const subCache = new Cache(this.controller, this.pageSize, 0, this, index);
+    this.#subCacheByIndex[index] = subCache;
+    return subCache;
+  }
+
+  /**
+   * Retrieves the flattened index corresponding to the given index
+   * of an item in the `items` array.
+   *
+   * @param {number} index
+   * @return {number}
+   */
+  getFlatIndex(index) {
```

Comment:
**note:** a copy-paste of the original getFlatIndex:

https://github.com/vaadin/web-components/blob/f1cea1bcb1d4bd2fe1f16390c452e2b2131823b2/packages/grid/src/vaadin-grid-data-provider-mixin.js#L107-L113